text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,585 @@
+                            // the values(feature) list
+fn write_shakespear() {}
+```
+
+## Cargo support
+
+Cargo is ideally positioned to enable checking for `feature` flags, since Cargo knows the set of
+valid features. Cargo will invoke `rustc --check-cfg 'values(feature, "...", ...)'`, so that
+checking for features is enabled. Optionally, Cargo could also specify the set of valid condition
+names.
+
+Cargo users will not need to do anything to take advantage of this feature. Cargo will always
+specify the set of valid `feature` flags. This may cause warnings in crates that contain invalid
+`#[cfg]` conditions. (Rust is permitted to add new lints; new lints are not considered a breaking
+change.) If a user upgrades to a version of Cargo / Rust that supports validating features, and
+their crate now reports errors, then they will need to align their source code with their
+`Cargo.toml` file in order to fix the error. (Or use `#[allow(...)]` to suppress it.) This is a
+benefit, because it exposes potential existing bugs.
+
+## Supporting build systems other than Cargo
+
+Some users invoke `rustc` using build systems other than Cargo. In this case, `rustc` will provide
+the mechanism for validating conditions, but those build systems will need to be updated in order
+to take advantage of this feature. Doing so is expected to be easy and non-disruptive, since this
+feature does not change the meaning of the existing `--cfg` option.
+
+# Reference-level explanation
+
+## What Cargo does
+
+When Cargo builds a `rustc` command line, it knows which features are enabled and which are
+disabled. Cargo normally specifies the set of enabled features like so:
+
+```bash
+rustc --cfg 'feature="lighting"' --cfg 'feature="bump_maps"' ...
+```
+
+When conditional compilation checking is enabled, Cargo will also specify which features are
+valid, so that `rustc` can validate conditional compilation tests. For example:
+
+```bash
+rustc --cfg 'feature="lighting"' --cfg 'feature="bump_maps"' \
+      --check-cfg 'values(feature, "lighting", "bump_maps", "mip_maps", "vulkan")'
+```
+
+In this command-line, Cargo has specified the full set of _valid_ features (`lighting`,
+`bump_maps`, `mip_maps`, `vulkan`) while also specifying which of those features are currently
+_enabled_ (`lighting`, `bump_maps`).
+
+## Command line arguments reference
+
+`rustc` accepts the `--check-cfg` option, which specifies whether to check conditions and how to
+check them. The `--check-cfg` option takes a value, called the _check cfg specification_. The
+check cfg specification is parsed using the Rust metadata syntax, just as the `--cfg` option is.
+(This allows for easy future extensibility, and for easily specifying moderately-complex data.)
+
+Each `--check-cfg` option can take one of two forms:
+
+1. `--check-cfg names(...)` enables checking condition names.
+2. `--check-cfg values(...)` enables checking the values within list-valued conditions.

```

Comment:
Nit: `--check-cfg-names=name1,name2,...`/`--check-cfg-values=name=value1,value2,...` would be more consistent with how the CLI hebaves in other areas of rustc. It'd be easier to implement in rustc and use by users of rustc too, I think.

Reply:
I considered that, but there are several reasons that I spec'd this using a single `--cfg` option. The most important reason is that the nature of checking `#[cfg]` may change over time, and may require expressing even more interesting or complex rules. It's fairly easy to express that kind of information in attribute / metadata syntax. If we had to add more complex information, then that requires adding more and more command-line parameters.

There is some complexity cost in carrying those command-line parameters through different layers of the compiler. I think having a single `--check-cfg` option has less complexity impact on the compiler implementation. It's just simpler, from the point of view of the command-line.